modules/edu.gemini.seqexec.engine/src/main/scala/edu/gemini/seqexec/engine/package.scala

```diff
@@ -22,7 +23,52 @@ package engine {
     def fromProcess(p: Process[Task, Event]): HandleP[Unit] = HandleP(Applicative[Handle].pure[(Unit, Option[Process[Task, Event]])](((), Some(p))))
   }
   final case class ActionMetadata(conditions: Conditions, operator: Option[Operator], observer: Option[Observer])
-  final case class Action(kind: ActionType, gen: Kleisli[Task, ActionMetadata, Result])
+  object ActionMetadata {
+    val default: ActionMetadata = ActionMetadata(Conditions.default, None, None)
+  }
+
+  final case class Action(
+    kind: ActionType,
+    gen: ActionGen,
+    state: Action.ActionState
+  )
+  object Action {
+
+    sealed trait ActionState
+
+    object ActionState {
+      implicit val equal: Equal[ActionState] = Equal.equalA
+    }
+
+    object Idle extends ActionState
+
+    object Started extends ActionState
+
+    final case class PartiallyCompleted[V <: PartialVal](pr: V) extends ActionState
+
+    object Paused extends ActionState
+
+    final case class Completed[V <: RetVal](r: V) extends ActionState
+
+    final case class Failed(e: Error) extends ActionState
+
+    def errored(ar: Action): Boolean = ar.state match {
+      case Action.Failed(_) => true
+      case _ => false
+    }
+
+    def finished(ar: Action): Boolean = ar.state match {
+      case Action.Failed(_) => true
```

Comment:
Is it so that it's finished if it failed?

Reply:
Yes. In this context, `finished` means "It was running, but it is not running anymore." It is different from completed.